tests/profile.py

```diff
@@ -0,0 +1,584 @@
+
+    dir_parser.add_argument(
+        "dirs",
+        nargs="+",
+        help="One or more directories to read files from (recursively)",
+    )
+
+    dir_parser.set_defaults(func=run_dir_profiler)
+
+    return parser.parse_args()
+
+
+def cast_to_bool(string):
+    '''
+    Ensure that the argument provided is either "True" or "False (or "true" or
+    "false") and convert that argument to an actual bool value (True or False).
+    '''
+    string_lower = string.lower()
+    if string_lower == "true":
+        return True
+    elif string_lower == "false":
+        return False
+    raise argparse.ArgumentTypeError('Argument must be True or False')
+
+
+def run_job_profiler(args):
+
+    # ------------------------------------------------
+    # Fetch Job
+    # ------------------------------------------------
+    logger.info("Fetching Job %s", args.jid)
+    client = api_client.ApiClient()
+    r_body, _ = client.make_request('api/v1/jobs',
+                                    params="filter=jid_eq_%s" % args.jid,
+                                    use_api_key=True
+                                    )
+
+    jobs = json.loads(r_body).get("data")
+    if not jobs:
+        raise Exception("No Job %s found" % args.jid)
+    assert len(jobs) < 2, "Found more than 1 Job with jid %s" % args.jid
+    job = jobs[0]
+    upload_id = job["upload"]
+
+    # ------------------------------------------------
+    # Fetch Job's Upload
+    # ------------------------------------------------
+    logger.info("Fetching Upload %s", upload_id)
+    r_body, _ = client.make_request('api/v1/uploads/%s' % upload_id,
+                                    use_api_key=True
+                                    )
+    upload = json.loads(r_body).get("data")
+    filepaths = upload["upload_files"].keys()
+
+    profile(
+        filepaths,
+        stat=args.stat,
+        read=args.read,
+        hash_md5=args.md5,
+        hash_xx=args.xxhash if XXHASH else None,
+        warmup=args.warmup,
+        read_size=args.read_size,
+        thread_count=args.threads,
+        csv_filepath=args.csv_path,
+    )
+
+
+def run_dir_profiler(args):
+    logger.info("Testing files in the following directories:\n\t%s", "\n\t".join(args.dirs))
+    # Validate that provided directories exist
+    for dirpath in args.dirs:
+        if not os.path.isdir(dirpath):
+            raise Exception("Directory does not exist: %s" % dirpath)
+
+    filepaths = []
+    for dirpath in args.dirs:
+        filepaths.extend(get_files(dirpath, recurse=True))
+
+    profile(
+        filepaths,
+        stat=args.stat,
+        read=args.read,
+        hash_md5=args.md5,
+        hash_xx=args.xxhash if XXHASH else None,
+        warmup=args.warmup,
+        read_size=args.read_size,
+        thread_count=args.threads,
+        csv_filepath=args.csv_path,
+        skip_failures=args.skip_failures,
+    )
+
+
+def profile(filepaths, stat=True, read=True, hash_md5=True, hash_xx=True, warmup=True,
```

Comment:
Can't we just use the default values defined for every argument? It feels redundant to redefine a second set of defaults here.

Reply:
This function was designed to stand on its own, and therefore provide its own default values. The fact that our shell/argparse logic happens to call it is irrelevant and should not dictate whether this function should define its own default arguments IMO.  For example, if I wanted to import/reuse this function elsewhere, I would want this function to provide useful default arguments.